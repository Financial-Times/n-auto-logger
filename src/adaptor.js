import logAction from './action';
import logOperation from './operation';

const adaptor = targetFunction => {
	if (targetFunction.length >= 2) {
		return logOperation(targetFunction);
	}
	if (targetFunction.length === 1) {
		return logAction(targetFunction);
	}
	throw Error(
		`targetFunction ${
			targetFunction.name
		} can not be decorated with autoLogger`,
	);
};

export default adaptor;
