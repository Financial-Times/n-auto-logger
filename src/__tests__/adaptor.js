import logger from '../index';
import adaptor from '../adaptor';

jest.mock('@financial-times/n-logger');

describe('adaptor', () => {
	afterEach(() => {
		jest.resetAllMocks();
	});

	it('recognise and enhance action function correctly', () => {
		const actionFunction = ({ paramA, meta }) => ({ paramA, ...meta });
		const enhanced = adaptor(actionFunction);
		const paramA = 'a';
		const meta = { userId: 'mock-id' };
		const result = enhanced({ paramA, meta });
		expect(result).toMatchSnapshot();
		expect(logger.info.mock.calls).toMatchSnapshot();
	});

	it('recognise and enhance operation function correctly', async () => {
		const operationFunction = (meta, req) => {
			req.meta = meta;
		};
		const enhanced = adaptor(operationFunction);
		const meta = { userId: 'mock-id' };
		const req = {};
		await enhanced(meta, req);
		expect(req.meta).toMatchSnapshot();
		expect(logger.info.mock.calls).toMatchSnapshot();
	});
});
