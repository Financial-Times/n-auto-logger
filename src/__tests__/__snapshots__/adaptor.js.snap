// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adaptor recognise and enhance action function correctly 1`] = `
Object {
  "paramA": "a",
  "userId": "mock-id",
}
`;

exports[`adaptor recognise and enhance action function correctly 2`] = `
Array [
  Array [
    Object {
      "action": "actionFunction",
      "meta": Object {
        "userId": "mock-id",
      },
      "paramA": "a",
    },
  ],
  Array [
    Object {
      "action": "actionFunction",
      "meta": Object {
        "userId": "mock-id",
      },
      "paramA": "a",
      "result": "success",
    },
  ],
]
`;

exports[`adaptor recognise and enhance operation function correctly 1`] = `
Object {
  "operation": "operationFunction",
  "userId": "mock-id",
}
`;

exports[`adaptor recognise and enhance operation function correctly 2`] = `
Array [
  Array [
    Object {
      "operation": "operationFunction",
      "userId": "mock-id",
    },
  ],
  Array [
    Object {
      "operation": "operationFunction",
      "result": "success",
      "userId": "mock-id",
    },
  ],
]
`;
